text/0000-time-improvements.md

```diff
@@ -0,0 +1,327 @@
+relaxed over time with additional types or through further RFCs:
+
+It provides convenience methods for constructing Durations from larger units
+of time (minutes, hours, days), but gives them names like
+`Duration.from_standard_hour`. A standard hour is 3600 seconds, regardless of
+
+It provides APIs that are expected to produce positive `Duration`s, and expects
+that APIs like timeouts will accept positive `Durations` (which is currently
+the case in Rust's standard library). These APIs help the programmer discover
+the possibility of system clock time travel, and either handle the error explicitly,
+or at least avoid propagating the problem into other APIs (by using `unwrap`).
+
+It separates monotonic time (`ProcessTime`) from time derived from the system
+clock (`SystemTime`), which must account for the possibility of time travel.
+This allows methods related to monotonic time to be uncaveated, while working
+with the system clock has more methods that return `Result`s.
+
+This RFC does not attempt to define a type for calendared DateTimes, nor does it
+directly address time zones.
+
+# Proposal
+
+## Types
+
+```rs
+pub struct ProcessTime {
+  secs: u64,
+  nanos: u32
+}
+
+pub struct SystemTime {
+  secs: u64,
+  nanos: u32
+}
+
+pub struct Duration {
+  secs: u64,
+  nanos: u32
+}
+```
+
+### ProcessTime
+
+`ProcessTime` is the simplest of the instant types. It represents an opaque
+(non-serializable!) timestamp that is guaranteed to be monotonic throughout
+the timeframe of the process it was created in.
+
+> In this context, monotonic means that a timestamp created later in real-world
+> time will always be larger than a timestamp created earlier in real-world
+> time.
+
+The `Duration` type can be used in conjunction with `ProcessTime`, and these
+operations have none of the usual time-related caveats.
+
+* Add a `Duration` to a `ProcessTime`, producing a new `ProcessTime`
+* compare two `ProcessTime`s to each other
+* subtract a `ProcessTime` from a later `ProcessTime`, producing a `Duration`
+* ask for an amount of time elapsed since a `ProcessTime`, producing a `Duration`
+
+Asking for an amount of time elapsed from a given `SystemTime` is a very common
+operation that is guaranteed to produce a positive `Duration`. Asking for the
+difference between an earlier and a later `SystemTime` also produces a positive
+`Duration` when used correctly.
+
+This design does not assume that negative `Duration`s are never useful, but
+rather than the most common uses of `Duration` do not have a meaningful
+use for negative values. Rather than require each API that takes a `Duration`
+to produce an `Err` (or `panic!`) when receiving a negative value, this design
+optimizes for the broadly useful positive `Duration`.
+
+```rs
+impl SystemTime {
+  /// Panics if `earlier` is later than &self.
+  /// Because SystemTime is monotonic, the only ime that `earlier` should be
+  /// a later time is a bug in your code.
```

Comment:
Are we going to guarantee that the time will "never" overflow? Windows guarantees that overflow won't happen for at least ~200 years of uptime, but I'm not sure about OSX and Linux.